044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,178 @@
+means the nodes will be able to communicate even if they are currently marked as not ready.
+
+### Impact on the KafkaRoller
+
+The existing KafkaRoller checks whether a pod is marked as ready to determine whether it needs to be rolled.
+This relies on the fact that currently the pod being ready implies that the BrokerState is >= RUNNING.
+This means any change to the readiness probes also impacts the KafkaRoller.
+In future it would be better for the KafkaRoller to do its own checks, rather than relying on the readiness probes implementation.
+
+The details for how the KafkaRoller should work going forwards should be discussed in a separate proposal.
+This section describes some ideas for how the KafkaRoller should work in KRaft mode based on investigations done while writing this proposal.
+It can be used as a starting point for the KafkaRoller in KRaft mode proposal.
+The key requirement for this proposal to be fully implemented is that the KafkaRoller takes into account the state of controller pods when deciding whether to roll broker pods.
+
+Based on investigations as part of writing this proposal, in KRaft mode:
+* The KafkaRoller should observe the BrokerState metric and any other metrics it needs directly, rather than inferring the state based on readiness.
+* If more than one controller pod has not become ready, KafkaRoller should try to roll the controllers that aren't the current active controller first.
+* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
+* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
+(i.e. not in pending state) before rolling the pod or waiting for it to become ready.
+
+The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
+The brokers do not move to RUNNING until a quorum has been formed.
+For example in combined mode during normal startup the following would happen:
+
+* All the pods are started at the same time
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to RUNNING and are marked as ready
+* All pods are now ready
+
+However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:
+
+* The first pod is scheduled
+* The controller process starts up but cannot form a quorum because the other controllers are missing
+* The broker process starts up but since there is no controller quorum it does not move to RUNNING, it stays in STARTING. This means the pod is not marked as ready
+* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff
+
+## Proposed order of tasks to complete this proposal
```

Comment:
These 3 phases are now describing the "listening" mentioning the ports directly (9090 and 9091) and not what's in the `inter.broker.listener.name` and `controller.listener.names`. Would it be better having consistency?

Reply:
I've edited the part higher up to refer to the ports first and then mention they are connected to those specific configurations. I think here it's fine to just refer to the ports as that is what will ultimately be in the checks.